app/screens/ScorecardPreference/ScorecardPreference.js

```diff
@@ -3,49 +3,103 @@ import {View, Text, StyleSheet, ScrollView, TouchableWithoutFeedback} from 'reac
 import MaterialIcon from 'react-native-vector-icons/MaterialIcons';
 import DatePicker from 'react-native-datepicker';
 import Moment from 'moment';
+import NetInfo from '@react-native-community/netinfo';
 
 import {LocalizationContext} from '../../components/Translations';
 import SelectPicker from '../../components/SelectPicker';
-import MessageLabel from '../../components/MessageLabel';
 import BottomButton from '../../components/BottomButton';
+import DownloadButton from '../../components/ScorecardDetail/DownloadButton';
 import HeaderTitle from '../../components/HeaderTitle';
 import ProgressHeader from '../../components/ProgressHeader';
+import ErrorMessageModal from '../../components/ErrorMessageModal/ErrorMessageModal';
+
+import { ERROR_INTERNET } from '../../constants/error_constant';
+
 import Color from '../../themes/color';
+
 import { getAll as getAllProgramLanguage } from '../../services/program_language_service';
+import {
+  save as saveProgramLanguage,
+  isExist as isProgramLanguageExist,
+} from '../../services/program_language_service';
+
 import scorecardService from '../../services/scorecardService';
+import authenticationService from '../../services/authentication_service';
+import { getErrorType } from '../../services/api_service';
+
 import { isKhmerLanguage } from '../../utils/program_language_util';
 
+import {
+  find as findScorecardDownload,
+  isDownloaded as isScorecardDownloaded,
+  download,
+  stopDownload,
+  getDownloadPercentage,
+  isAudioLanguageDownloaded,
+  downloadLangIndicatorAudio,
+} from '../../services/scorecard_download_service';
+
 class ScorecardPreference extends Component {
   static contextType = LocalizationContext;
   constructor(props) {
     super(props);
     this.state = {
-      scorecard: '',
+      scorecard: scorecardService.find(this.props.route.params.scorecard_uuid),
       languages: [],
       textLocale: '',
       audioLocale: '',
       date: Moment().format('DD/MM/YYYY'),
-      message: '',
-      messageType: '',
+      downloadProgress: 0,
+      showDownloadProgress: false,
+      isStopDownload: false,
+      visibleModal: false,
+      errorType: null,
+      isFinishDownloaded: false,
+      isDownloading: false,
+      isErrorDownload: false,
+      isAudioDownloaded: false,
+      isDownloadAudio: false,
+      hasInternetConnection: false,
     };
     this.textLanguageController;
     this.audioLanguageController;
   }
 
   componentDidMount() {
-    const scorecard = scorecardService.find(this.props.route.params.scorecard_uuid);
-    this.setState({scorecard: scorecard}, () => {
-      this.loadProgramLanguage();
+    NetInfo.addEventListener((state) => {
+      if (state.isInternetReachable != null && state.isInternetReachable)
```

Comment:
@limkimsan consider to remove this into a function(hasInternetConnection()) which return true/false -> then this.setState({hasInternetConnection: hasInternetConnection()}